### ISSUE TITLE
Problem with KCL ArgoCD plugin

### ISSUE THREAD

**Issue by wilsonwang371** [kcl-lang/kcl#1162]
## General Question

I tried to exec `kcl run` in the repo and it is fine. But when in ArgoCD plugin container, I got error
```
failed to compile the kcl package unexpected EOF
```

what can be the root cause of this?

**Comment by wilsonwang371** [comment]
I think I found the issue. It was because downloading k8s package failed. Our enterprise environment has restricted access to public registries, maybe.

**Comment by Peefy** [comment]
> I think I found the issue. It was because downloading k8s package failed. Our enterprise environment has restricted access to public registries, maybe.

I see. Perhaps you can clone packages such as `k8s` to the internal git repository or maintain them in the internal registry.

**Comment by wilsonwang371** [comment]
> > I think I found the issue. It was because downloading k8s package failed. Our enterprise environment has restricted access to public registries, maybe.
> 
> I see. Perhaps you can clone packages such as `k8s` to the internal git repository or maintain them in the internal registry.

How do I change the default registry? Or anyway I pre-download the packages to a KCL system wide directory?

**Comment by Peefy** [comment]
> > > I think I found the issue. It was because downloading k8s package failed. Our enterprise environment has restricted access to public registries, maybe.
> > 
> > 
> > I see. Perhaps you can clone packages such as `k8s` to the internal git repository or maintain them in the internal registry.
> 
> How do I change the default registry? Or anyway I pre-download the packages to a KCL system wide directory?

See [here](https://www.kcl-lang.io/docs/user_docs/guides/package-management/how-to/kpm_oci)

**Comment by wilsonwang371** [comment]
> > > > I think I found the issue. It was because downloading k8s package failed. Our enterprise environment has restricted access to public registries, maybe.
> > > 
> > > 
> > > I see. Perhaps you can clone packages such as `k8s` to the internal git repository or maintain them in the internal registry.
> > 
> > 
> > How do I change the default registry? Or anyway I pre-download the packages to a KCL system wide directory?
> 
> See [here](https://www.kcl-lang.io/docs/user_docs/guides/package-management/how-to/kpm_oci)

How do I download packages to local KPM directory? If I pre download the ones I needed will it work?

**Comment by Peefy** [comment]
You can use `kcl mod add path/to/local/module` to add this module into your kcl program.

**Comment by Peefy** [comment]
cc @zong-zhe Could you please add this internal module guide to the website?

**Comment by wilsonwang371** [comment]
> You can use `kcl mod add path/to/local/module` to add this module into your kcl program.

What if A import B and B imports a local relative path? I saw error:

```
failed to compile the kcl package
failed to hash the kcl package 'k8s' in '3rdparty/ghcr.io/kcl-lang/k8s'.
lstat 3rdparty/ghcr.io/kcl-lang/k8s: no such file or directory
```

**Comment by Peefy** [comment]
Hello. What do you mean is that running `A` will result in errors? How did `A` import `B`?

**Comment by wilsonwang371** [comment]
A imports B where B is at a private gitlab repo. B imports k8s from the repo local directory. When I run A, I saw such issue.

**Comment by Peefy** [comment]
Make sense. @zong-zhe Can you help determine if this is a bug?

**Comment by zong-zhe** [comment]
Hi @wilsonwang371 

On which platform did you have this problem?  linux or macos ?

If this path `3rdparty/ghcr.io/kcl-lang/k8s` exists, make sure **it is a file path and not a symbolic link**. kpm will download the dependency and add a symbolic link to the packages downloaded in the latest time period, which will be used as the default dependency for the `*.k` files that do not have `kcl.mod`.

**Comment by wilsonwang371** [comment]
> Hi @wilsonwang371 
> 
> On which platform did you have this problem?  linux or macos ?
> 
> If this path `3rdparty/ghcr.io/kcl-lang/k8s` exists, make sure **it is a file path and not a symbolic link**. kpm will download the dependency and add a symbolic link to the packages downloaded in the latest time period, which will be used as the default dependency for the `*.k` files that do not have `kcl.mod`.

on macos. but it might also be there on linux. it is not a symlink.

**Comment by wilsonwang371** [comment]
@Peefy are we going to make a release after this?

**Comment by Peefy** [comment]
Hello @wilsonwang371 This issue is expected to be fixed today (another PR is needed), and I will release version 0.8.3 and the latest image at that time.

**Comment by wilsonwang371** [comment]
> Hello @wilsonwang371 This issue is expected to be fixed today (another PR is needed), and I will release version 0.8.3 and the latest image at that time.

when will 0.8.3 be available? thanks

**Comment by Peefy** [comment]
> when will 0.8.3 be available? thanks

Today.

**Comment by Peefy** [comment]
Hello @wilsonwang371  I've release the KCL v0.8.3 and the latest `kcllang/kcl` image. Thank you for the feedback again.

**Comment by wilsonwang371** [comment]
btw, there is another issue. After execution, the kcl.mod is updated with an absolute path. This local environment location should not be updated, right?


![image](https://github.com/kcl-lang/kcl/assets/3913185/eac97d3c-f391-4b53-afcd-0acfe84d48fb)

**Comment by wilsonwang371** [comment]
another issue is that, package B depends on package A. Package A depends on package k8s. In this case after I run kcl. I saw that the absolute path of k8s is also updated to package B. I don't think this is what users want during development. 

![image](https://github.com/kcl-lang/kcl/assets/3913185/20315777-7ae5-46a7-98f3-b5f95a0a847b)

**Comment by Peefy** [comment]
cc @zong-zhe

**Comment by zong-zhe** [comment]
Hi, @wilsonwang371 😄

The phenomenon you are observing is primarily caused by two parts.

- Part One: The current design of kpm updates all dependencies on the entire dependency graph and flattens them into the `kcl.mod` file at the build entry point. For example, if you have a dependency chain of `A -> B -> C`, you will see both dependencies `B` and `C` in the `kcl.mod` of `A`. This approach is similar to how `go.mod` works. Furthermore, the entire chain will use the dependencies defined in A's `kcl.mod`. The advantage of this method is the flexibility it offers in adjusting dependencies. For instance, if the official release of package `B` depends on version 0.1.0 of `C`, but you actually need to use version 0.1.2 of `C` to fit your infrastructure, you do not need to ask the `B` package for update the version of `C`. Instead, you can make changes directly in the `kcl.mod`. Therefore, the `k8s` dependency from `kfoundations` is also updated into the `kcl.mod` file of your current KCL package, which serves as the build entry point.

- Part Two: Since the relative path of `k8s` is relative to `kfoundations`, when this dependency is updated in the `kcl.mod` file at the current build entry point, the relative path needs to be recalculated. And I am currently researching and designing the addressing method for relative paths.

Currently, I have two ideas:
* kpm could directly compute the path of C relative to A. This could result in the kcl.mod becoming quite unsightly after the update, especially if the dependency chain is long; you might end up seeing paths like `././.././././.././././../../././k8s`
* Introduce a variable in the path to indicate the starting point of the relative path, similar to `{B}/./C`, which means to find `C` relative to the `B` package in the current package.

These are just two simple ideas. I will continue to update my research and design later today.

**Comment by wilsonwang371** [comment]
> Hi, @wilsonwang371 😄
> 
> The phenomenon you are observing is primarily caused by two parts.
> 
> - Part One: The current design of kpm updates all dependencies on the entire dependency graph and flattens them into the `kcl.mod` file at the build entry point. For example, if you have a dependency chain of `A -> B -> C`, you will see both dependencies `B` and `C` in the `kcl.mod` of `A`. This approach is similar to how `go.mod` works. Furthermore, the entire chain will use the dependencies defined in A's `kcl.mod`. The advantage of this method is the flexibility it offers in adjusting dependencies. For instance, if the official release of package `B` depends on version 0.1.0 of `C`, but you actually need to use version 0.1.2 of `C` to fit your infrastructure, you do not need to ask the `B` package for update the version of `C`. Instead, you can make changes directly in the `kcl.mod`. Therefore, the `k8s` dependency from `kfoundations` is also updated into the `kcl.mod` file of your current KCL package, which serves as the build entry point.
> 
> - Part Two: Since the relative path of `k8s` is relative to `kfoundations`, when this dependency is updated in the `kcl.mod` file at the current build entry point, the relative path needs to be recalculated. And I am currently researching and designing the addressing method for relative paths.
> 
> I will continue to fix this issue today.
> 
> 

r we able to use a cache mechanism(a .cache directory for temp files) to solve part2?

**Comment by zong-zhe** [comment]
Hi @wilsonwang371 😄

https://github.com/kcl-lang/kpm/pull/291 has remove the indirect dependencies in kcl.mod.

**Comment by wilsonwang371** [comment]
> Hi @wilsonwang371 😄
> 
> [kcl-lang/kpm#291](https://github.com/kcl-lang/kpm/pull/291) has remove the indirect dependencies in kcl.mod and v0.8.4 is released.

Using 0.8.4, I saw that for my case (B depends on A and A depends on k8s), the first time running `kcl run` it will execute successfully, but the second time when I run `kcl run` I saw error saying k8s cannot be found.

**Comment by wilsonwang371** [comment]
BTW, In repo B's kcl.mod.lock, the k8s is never added to the dependencies.

**Comment by wilsonwang371** [comment]
I manually tried to clean B's kcl.mod.lock, then execute `kcl run` several times, each time I saw the content of kcl.mod.lock

1. Empty 
![image](https://github.com/kcl-lang/kcl/assets/3913185/55d43572-299d-4b78-b5c6-2968fc49903e)

2. run `kcl run` 
![image](https://github.com/kcl-lang/kcl/assets/3913185/19891425-e04b-4055-bf57-a7fe50932102)

3.  run `kcl run` again. Error happens 
![image](https://github.com/kcl-lang/kcl/assets/3913185/16186ef4-55ee-4033-a75b-d9a22059e2d9)

**Comment by Peefy** [comment]
cc @zong-zhe Please help investigate and fix it

**Comment by wilsonwang371** [comment]
@zong-zhe is this fixed in latest main? btw, when will next release be available? @Peefy

**Comment by Peefy** [comment]
Hello @wilsonwang371 Sorry for the delay, there are still some minor issues that need to be fixed and more e2e tests added. It is expected to be completed today.

**Comment by Peefy** [comment]
Hello @wilsonwang371 We've released v0.8.5. We still have some other optimizations that have not been completed, which may occur in v0.8.6.

**Comment by wilsonwang371** [comment]
@Peefy any update on this？

**Comment by Peefy** [comment]
> @Peefy any update on this？

Hello @wilsonwang371  Not currently available. The latest KCL version is v0.8.5, and other updates or fixes are not currently related to this issue. cc @zong-zhe

**Comment by zong-zhe** [comment]
Hi @wilsonwang371 
We've recently discovered some performance problem that need fixing, and you can track the progress of that in issue https://github.com/kcl-lang/kpm/issues/295.
